cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java

```diff
@@ -0,0 +1,136 @@
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.internal.app.runtime.schedule.TriggeringScheduleInfoAdapter;
+import co.cask.cdap.internal.app.store.RunRecordMeta;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Heartbeat Store that writes heart beat messages and program status messages
+ * to program heartbeat table. This is used for efficiently
+ * scanning and returning results for dashboard status queries
+ */
+public class ProgramHeartbeatDataset extends AbstractDataset {
+  private static final Gson GSON = TriggeringScheduleInfoAdapter.addTypeAdapters(new GsonBuilder()).create();
+  private static final byte[] COLUMN_NAME = Bytes.toBytes("status");
+  private static final DatasetId PROGRAM_HEARTBEAT_INSTANCE_ID =
+    NamespaceId.SYSTEM.dataset(Constants.ProgramHeartbeat.TABLE);
+  private static final byte[] ROW_KEY_SEPARATOR = Bytes.toBytes(":");
+
+  private final Table table;
+
+  public ProgramHeartbeatDataset(Table table) {
+    super("ignored", table);
+    this.table = table;
+  }
+
+  /**
+   * Static method for creating an instance of {@link ProgramHeartbeatDataset}.
+   */
+  public static ProgramHeartbeatDataset getOrCreate(DatasetContext datasetContext, DatasetFramework datasetFramework) {
+    try {
+      Table table = DatasetsUtil.getOrCreateDataset(datasetContext, datasetFramework, PROGRAM_HEARTBEAT_INSTANCE_ID,
+                                                    Table.class.getName(), DatasetProperties.EMPTY);
+      return new ProgramHeartbeatDataset(table);
+    } catch (DatasetManagementException | IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Write {@link RunRecordMeta} to heart beat dataset as value,
+   * with rowKey being <timestamp-in-seconds>:<program-run-id>
+   * @param runRecordMeta row value to write
+   * @param timestampInSeconds used for creating rowKey
+   */
+  public void writeRunRecordMeta(RunRecordMeta runRecordMeta, long timestampInSeconds) {
+    byte[] rowKey = createRowKey(timestampInSeconds, runRecordMeta.getProgramRunId());
+    table.put(rowKey, COLUMN_NAME, Bytes.toBytes(GSON.toJson(runRecordMeta)));
+  }
+
+  private byte[] createRowKey(long timestamp, ProgramRunId programRunId) {
+    return Bytes.add(Bytes.toBytes(timestamp), ROW_KEY_SEPARATOR, Bytes.toBytes(GSON.toJson(programRunId)));
```

Comment:
since a long is always a set number of bytes, you don't need the separator.

Reply:
having json in the rowkey seems like a bad idea. For example, if variables in the class get re-ordered, the row keys will change. Is there a reason we don't want to use the MDSKey that other datasets use?

Reply:
A table like this will also be subject to hotspotting since it's only every writing newer timestamps. Is there a plan to salt this table?

Reply:
good point, i will remove the separator. About salting, i wanted to add salting, however i realized our use case is read heavy than write heavy, writes from running programs every 30min vs UI making requests from dashboard.  Salting would be good for a write heavy workload. would like to hear about your thoughts. it also reminds me that i should add TTL for this table.

Reply:
I initially looked at `MDSKey` and after seeing the javadoc which says `rowkey for metadata` i skipped it, however after looking at its implementation its not specific to metadata and ideally should be named `RowKey` since now we are using it at multiple places.